File: back-end-java/src/main/java/com/aie/backend/modules/govbr/controller/GovbrController.java
package com.aie.backend.modules.govbr.controller;

import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.view.RedirectView;

import com.aie.backend.modules.govbr.service.GovbrService;

@RestController
@Validated
public class GovbrController {

    private final GovbrService service;

    public GovbrController(GovbrService service) {
        this.service = service;
    }

    @GetMapping("/govbr/authorize")
    public RedirectView authorize(@RequestParam(value = "origin", required = false) String origin) {
        String url = service.buildAuthorizeUrl(origin);
        return new RedirectView(url);
    }

    @GetMapping("/govbr/callback")
    public ResponseEntity<String> callbackLegacy(
            @RequestParam(value = "code", required = false) String code,
            @RequestParam(value = "state", required = false) String state,
            @RequestParam(value = "error", required = false) String error,
            @RequestParam(value = "error_description", required = false) String errorDescription
    ) {
        return callbackInternal(code, state, error, errorDescription);
    }

    @GetMapping("/retornoWebHook")
    public ResponseEntity<String> callback(
            @RequestParam(value = "code", required = false) String code,
            @RequestParam(value = "state", required = false) String state,
            @RequestParam(value = "error", required = false) String error,
            @RequestParam(value = "error_description", required = false) String errorDescription
    ) {
        return callbackInternal(code, state, error, errorDescription);
    }

    private ResponseEntity<String> callbackInternal(
            String code,
            String state,
            String error,
            String errorDescription
    ) {
        String origin = service.extractOriginFromState(state);
        if (error != null) {
            String html = service.renderPopupClosePage(origin, java.util.Map.of(
                    "status", "error",
                    "message", errorDescription != null ? errorDescription : error,
                    "error", error
            ));
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
                    .body(html);
        }

        try {
            var result = service.handleCallback(code, state);
            String html = service.renderPopupClosePage(
                    (String) result.get("origin"),
                    java.util.Map.of(
                            "status", "success",
                            "token", result.get("token"),
                            "user", result.get("user")
                    )
            );
            return ResponseEntity.ok()
                    .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
                    .body(html);
        } catch (Exception ex) {
            String message = ex.getMessage() != null ? ex.getMessage() : "Falha ao autenticar no gov.br";
            String html = service.renderPopupClosePage(origin, java.util.Map.of(
                    "status", "error",
                    "message", message,
                    "error", ex.getClass().getSimpleName()
            ));
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
                    .body(html);
        }
    }
}

File: back-end-nestjs/src/modules/govbr/govbr.controller.ts
import { Controller, Get, Query, Res } from '@nestjs/common';
import type { Response } from 'express';
import { GovbrAuthorizeQueryDto } from './dto/govbr-authorize-query.dto';
import { GovbrCallbackDto } from './dto/govbr-callback.dto';
import { GovbrService } from './govbr.service';
import { ApiInternalServerErrorResponse } from '@nestjs/swagger';
import { InternalServerError } from 'src/shared/config/swagger/responses/errors';

@ApiInternalServerErrorResponse(InternalServerError)
@Controller()
export class GovbrController {
  constructor(private readonly service: GovbrService) {}

  @Get('govbr/authorize')
  async authorize(
    @Query() query: GovbrAuthorizeQueryDto,
    @Res() res: Response,
  ) {
    try {
      const url = await this.service.buildAuthorizeUrl(query.origin);
      return res.redirect(url);
    } catch (err: any) {
      const html = this.service.renderPopupClosePage(
        this.service.getOriginForResponse(query.origin, undefined),
        {
          status: 'error',
          message:
            err?.message || 'Não foi possível iniciar o login com o gov.br',
          error: err?.name,
        },
      );
      return res.status(400).type('html').send(html);
    }
  }

  @Get('govbr/callback')
  async callbackLegacy(@Query() query: GovbrCallbackDto, @Res() res: Response) {
    return this.handleCallback(query, res);
  }

  @Get('retornoWebHook')
  async callback(@Query() query: GovbrCallbackDto, @Res() res: Response) {
    return this.handleCallback(query, res);
  }

  private async handleCallback(query: GovbrCallbackDto, res: Response) {
    const originFromState = this.service.extractOriginFromState(query.state);
    const origin = this.service.getOriginForResponse(
      undefined,
      originFromState ?? undefined,
    );

    if (query.error) {
      const html = this.service.renderPopupClosePage(origin, {
        status: 'error',
        message: query.error_description || query.error,
        error: query.error,
      });
      return res.status(400).type('html').send(html);
    }

    try {
      const result = await this.service.handleCallback(query.code, query.state);
      const html = this.service.renderPopupClosePage(result.origin, {
        status: 'success',
        token: result.token,
        user: result.user,
      });
      return res.type('html').send(html);
    } catch (err: any) {
      const html = this.service.renderPopupClosePage(origin, {
        status: 'error',
        message:
          err?.message ||
          'Não foi possível concluir o login com o provedor gov.br',
        error: err?.name,
      });
      return res.status(400).type('html').send(html);
    }
  }
}
